Workiva/over_react
# CPLAT-11722 Add uiForwardRef

## Motivation
  
When planning function components, we missed a couple cases:

- when you want to forward a ref to a function component
- when you want to forward a ref using something than the existing component's props (way less common / important to solve)
Example: https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-to-dom-components
## Changes
  
API Changes:
- Add `uiForwardRef`
- Update ref example to include different syntax examples
- Refactor forwardRef tests to be able to share test coverage with `uiForwardRef`
- Add `uiForwardRef` tests

Builder Changes:
- Rename `FunctionComponentConfig` to `UiFactoryConfig`
- Rename / refactor `isFunctionComponentFactory` to `shouldGenerateConfig`
- Create a getter on `Identifier` that can be used to check if the name is a factory type
- Switch out a couple checks on the identifier name with checks for the presence of a generated `UiFactoryConfig`

#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        
      - [ ] QA that the refs print correctly in the example. Note that the components do have code around logging, but that's to keep some consistency between React's docs and our examples. Because there aren't any props updated, there isn't any re-rendering and no reason for the components to log.

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
